# Part 1
## a. Formulating a knowledge base (KB) using propositional logic
### The given puzzle is:

1. Chen's mark was lower than the mark of the person who studied French.<p>
   

2. The three people are Brian, the person who studied History, the person who graduated on the 2nd.<p>
   

3. Chen scored 70%, unless she studied history in which case her mark was 65%.<p>
    

4. The mark of the person who graduated on the 3rd was 5% higher than Chen's. <p>
 

5. Anjali did not graduate on the 2nd.<p>
 

### Solution


a. Brian didn't study history or graduate on 2nd. (From 2)<p>


b. Chen graduated on 2nd. (From 1,2 and 5)<p>

c. Chen scored 70%. (From 3)<p>

d. Chen studied Geography. (From 1,2,3,b and c)<p>


e. Anjali studied History. (From 2 and 5) <p>
    

f. Anjali scored 65%. (From 3 and e) <p>

g. Anjali graduated on 4th. (From 4 and f) <p>

h. Brian studied French. (From d and e) <p>

i. Brian scored 75%. (From c and f) <p>

j. Brian graduated on 3rd. (From b and g)<p>

So for final solution, we have,

|Name  |Graduated on|Subject  |Score|
|------|------------|---------|-----|
|Chen  |2nd         |Geography|70%. |
|Brian |3rd         |French.  |75%. |
|Anjali|4th         |History. |65%. |

### Defining the propositions

AF: Anjali studied French, AG: Anjali studied Geography, AH: Anjali studied History.<p>
BF: Brian studied French, BG: Brian studied Geography, BH: Brian studied History.<p>
CF: Chen studied French, CG: Chen studied Geography, CH: Chen studied History.<p>
A2: Anjali graduated on 2nd, A3: Anjali graduated on 3rd, A4: Anjali graduated on 4th.<p>
B2: Brian graduated on 2nd, B3: Brian gradutaed on 3rd, B4: Brian graduated on 4th.<p>
C2: Chen graduated on 2nd, C3: Chen graduated on 3rd, C4: Chen graduated on 4th.<p>
A65: Anjali scored 65, A70: Anjali scored 70, A75: Anjali scored 75.<p>
B65: Brian scored 65, B70: Brian scored 70, B75: Brian scored 75.<p>
C65: Chen scored 65, C70: Chen scored 70, C75: Chen scored 75.<p>

### For knowledge base representation, we have

1. Chen's mark was lower than the mark of the person who studied French.<p>
    Clue 1: So Chen's mark is not 75 and he didnt study French.
    
      $ R1: \neg CF \land \neg C75 $<p>
    Clue 2: If Chen didnt study French, then she studied either Geography or History.
   
      $ R2: CG \lor CH$<p>
    Clue 3: We know Chen didnt study French. So either Anjali or Brian studied French. So, we have for Anjali studying French,
        
      $ R3: AF \implies\ (BG \lor BH) $<p>
        
        And for Brian studying French,
      $ R4: BF \implies\ (AG \lor AH) $<p>

2. The three people are Brian, the person who studied History, the person who graduated on the 2nd.<p>
    
    Clue 1: Since Brian didnt study history, he either studied Geography or French. Also, he didnt graduate on 2nd so he graduated on either 3rd or 4th.
    
    $ R5: ((BG \lor BF)$<p>
    $ R6: (B3 \lor B4)$ <p>
    $ R7: (\neg BH \land \neg B2))$<p> 
    $ R8: (AH \lor CH)$<p>
    $ R9: (C2 \lor A2)$<p>
        
    Clue 2: If Anjali studied History, we have,
        
    $ R10: AH \implies\ C2$<p>
    $ R11: AH \implies\ (A3 \lor A4)$<p>
    $ R12: AH \implies\ (CG \lor CF)$
    
    Clue 3: If Chen studied History, we have,
     
    $ R13: CH \implies\ (AG \lor AF)$<p>
    $ R14: CH \implies\ (C3 \lor C4)$
    
    Clue 4: If Chen graduated on 2nd then,
        
    $ R15: C2 \implies\ (A3 \lor A4)$
        
    Clue 5: For Anjali graduating on 2nd,   
        
    $ R16: A2 \implies\ (C3 \lor C4)$

3. Chen scored 70%, unless she studied history in which case her mark was 65%.<p>
    
    Clue 1: For Chen scoring 70, we have,
    
    R17: $ CG \implies\ C70 $<p>
    R18: $ CF \implies\ C70 $<p>
        
    We know Chen's mark is lower than the person who studied French. So,<p>
    
    R19: $ BF \implies\ B75$<p>
    R20: $ AF \implies\ A75$<p>
    
    Clue 2: For Chen studying history, we have,
        
    R21: $ CH \implies\ C65$<p>
    R22: $ CH \implies\ (BF \lor B70) $<p>
    R23: $ CH \implies\ (AF \lor A70)$
        
   

4. The mark of the person who graduated on the 3rd was 5% higher than Chen's. 
    
    Clue 1: Since the person who graduated on the 3rd was 5% higher than Chen chen couldnt have graduated on 3rd. Also he wouldnt have the highest score. And either Anjali or Brian graduated on 3rd. 
    
    R24: $\neg C3 \land \neg C75$
    R25: $ A3 \lor B3$
    
    Clue 2: Chen must have graduated on 2nd or 4th.
    
    R26: $C4 \lor C2$
    
    Clue 3: Since Chen's mark is 5% lower than Anjali or Brian's mark. We have,
    
    R27: $C70 \implies\ (A75 \lor B75)$
   

5. Anjali did not graduate on the 2nd.<p>
    Clue 1: This means she graduated on either 3rd or 4th.<p>
    R28: $\neg A2$
    R29: $(A3 \lor A4)$
     
    Clue 2: Since Anjali didnt graduate on 2nd. Either Brian or Chen graduate on 2nd.
        
    R30: $B2 \implies\ (C3 \lor C4)$<p>
    R31: $B2 \implies\ (A3 \lor A4)$<p>
    R32: $C2 \implies\ (B3 \lor B4)$<p>
    R33: $C2 \implies\ A4$

For 3 students with 3 subjects, 3 scores and 3 graduation dates, we have the expression as:

$f$(P,Q,R) = $(P \land \neg Q \land \neg R)\lor(\neg P \land Q \land \neg R)\lor(\neg P \land \neg Q \land R)$<p>
    This can be interpreted as:<p>
    $(P \lor Q \lor R) \land (\neg P \lor \neg Q) \land (\neg P \lor \neg R)\land (\neg Q \lor \neg R)$

6. Each student studied only one subject. <p>
    R34 : $f$(AF,AG,AH)<p>
    R35 : $f$(BF,BG,BH)<p>
    R36 : $f$(CF,CG,CH)<p>
                                         

7. Each student graduated on a different day.<p>
    R37 : $f$(A2,A3,A4)<p>
    R38: $f$(B2,B3,B4)<p>
    R39: $f$(C2,C3,C4)<p>
        

8. Each student scored different scores.<p>
    R40: $f$(A65,A70,A75)<p>
    R41: $f$(B65,B70,B75)<p>
    R42: $f$(C65,C70,C75)<p>

9. For individual subject,scores and marks, we have<p>
$R_{43}$: $f(AH, BH, CH)$ <p>
$R_{44}$: $f(AG, BG, CG)$<p>
$R_{45}$: $f(AF, BF, CF)$<p>
$R_{46}$: $f(A65, B65, C65)$<p>
$R_{47}$: $f(A70, B70, C70)$<p>
$R_{48}$: $f(A75, B75, C75)$<p>
$R_{49}$: $f(A2, B2, C2)$<p>
$R_{50}$: $f(A3, B3, C3)$<p>
$R_{51}$: $f(A4, B4, C4)$<p>

Adopting the relatively natural ordering
|AF|BF|CF|AG|BG|CG|AH|BH|CH|A2|B2|C2|A3|B3|C3|A4|B4|C4|A65|B65|C65|A70|B70|C70|A75|B75|C75|
|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|---|---|---|---|---|---|---|---|---|
|0 |1 |2 |3 |4 |5 |6 |7 |8 |9 |10|11|12|13|14|15|16|17|18 |19 |20 |21 |22 |23 |24 |25 |26 |

we now 

## b. Backtracking search

### First converting the constraints into Conjunctive Normal Form (CNF)

$1. \neg CF \land \neg C75$

$2. CG \lor CH$

$3. AF \implies\ (BG \lor BH)$<p>
    $ \neg AF \lor (BG \lor BH)$<p>

$4. BF \implies\ (AG \lor AH)$<p>
    $ \neg BF \lor (AG \lor AH)$

$5. \neg BH \lor \neg B2$

$6. BG \lor BF$

$7.B3 \lor B4$

$8. AH \lor CH$

$9. A2 \lor C2$

$10. AH \implies\ C2$<p>
    $ \neg AH \lor C2$

$11. AH \implies\ A3 \lor A4$<p>
    $ \neg AH \lor (A3 \lor A4)$

$12. AH \implies\ (CG \lor CF)$<p>
    $ \neg AH \lor (CG \lor CF)$

$13. CH \implies\ (AG \lor AF)$<p>
   $ \neg CH \lor (AG \lor AF)$

$14. CH \implies\ (C3 \lor C4)$<p>
    $ \neg CH \lor (C3 \lor C4)$

$15. C2 \implies\ (A3 \lor A4)$<p>
    $ \neg C2 \lor (A3 \lor A4)$

$16. A2 \implies\ (C3 \lor C4)$<p>
    $ \neg A2 \lor (C3 \lor C4)$

$17. CG \implies\ C70$<p>
$ \neg CG \lor C70$

$18. CF \implies\ C70$<p>
    $ \neg CF \lor C70$

$19. BF \implies\ B75 $<p>
    $ \neg BF \lor B75$

$20. AF \implies A75$<p>
    $ \neg AF \lor A75$

$21. CH \implies\ C65$<p>
    $\neg CH \lor C65$

$22. CH \implies\ (BF \lor B70)$<p>
$ \neg CH \lor (BF \lor B70)$

$23. CH \implies\ (AF \lor A70)$<p>
$ \neg CH \lor (AF \lor A70)$

$24. \neg C3 \land \neg C75$

$25. A3 \lor B3$

$26. C4 \lor C2$

$27. C70 \implies\ (A75 \lor B75)$<p>
    $ \neg C70 \lor (A75 \lor B75)$

$28. \neg A2$

$29. A3 \lor A4$

$30. B2 \implies\ (C3 \lor C4)$<p>
    $ \neg B2 \lor (C3 \lor C4)$

$31. B2 \implies\ (A3 \lor A4)$<p>
    $ \neg B2 \lor (A3 \lor A4)$

$32. C2 \implies\ (B3 \lor B4)$<p>
    $ \neg C2 \lor (B3 \lor B4)$

$.33. C2 \implies\ A4$<p>
    $ \neg C2 \lor A4$

$34. (P \land \neg Q \land \neg R) \lor (\neg P \land Q \land \neg R) \lor (\neg P \land \neg Q \land R)$<p> 
    $(P \lor Q \lor R) \land (\neg P \lor \neg Q) \land (\neg P \lor \neg R) \land (\neg Q \lor \neg R)$

### Since the above equation is repeated in 9 constraints, we take the values of the constraints in the equation.

In [2]:
def EvaluateKB(Clauses, Model):
    EvaluatedClauses = []
    for clause in Clauses:
        EvaluatedLiterals = []
        for literal in clause:
            if literal[0] == '~':
                negated_literal = literal[1:]
                if Model.get(negated_literal) is None:
                    EvaluatedLiterals.append(None)
                else:
                    EvaluatedLiterals.append(not Model[negated_literal])
            else:
                if Model.get(literal) is None:
                    EvaluatedLiterals.append(None)
                else:
                    EvaluatedLiterals.append(Model[literal])
        if True in EvaluatedLiterals:
            EvaluatedClauses.append(True)
        elif None in EvaluatedLiterals:
            EvaluatedClauses.append(None)
        else:
            EvaluatedClauses.append(False)

    if all(i is True for i in EvaluatedClauses):
        EvaluatedKB = True
    elif False in EvaluatedClauses:
        EvaluatedKB = False
    else:
        EvaluatedKB = None

    return EvaluatedKB

In [3]:
import copy

def BackTrackingSearch(Clauses, Model):
    EvaluatedClauses = EvaluateKB(Clauses, Model)
    if EvaluatedClauses is True:
        return True, Model
    elif EvaluatedClauses is False:
        return False, None
    else:
        Model2 = copy.deepcopy(Model)
        NextSymbol = [i for i in Model2 if Model2[i] is None][0]
        Model2[NextSymbol] = True
        ModelTrue, SATModel = BackTrackingSearch(Clauses, Model2)
        if ModelTrue:
            return True, SATModel
        Model2[NextSymbol] = False
        ModelFalse, SATModel = BackTrackingSearch(Clauses, Model2)
        if ModelFalse:
            return True, SATModel
    return False, None

In [4]:
Clauses = [
   #1
    ('~CF',),
    ('~C75',),#R1
    ('CG', 'CH'),#R2
    ('~AF','BG','BH'),#R3
    ('~BF','AG','AH'),#R4
    #2
    ('~BH', '~B2'),#R5
    ('BG','BF'),#R6
    ('B3','B4'),#R7
    ('AH','CH'),#R8
    ('A2','C2'),#R9
    ('~AH', 'C2'),#R10
    ('~AH','A3','A4'),#R11
    ('~AH','CG','CF'),#R12
    ('~CH','AG','AF'),#R13
    ('~CH','C3','C4'),#R14
    ('~C2', 'A3','A4'),#R15
    ('~A2','C3','C4'),#R16
   
    #3
    ('~CG','C70'),#R17
    ('~CF','C70'),#R18
    ('~BF','B75'),#R19
    ('~AF','A75'),#R20
    ('~CH','C65'),#R21
    ('~CH','BF','B70'),#R22
    ('~CH','AF','A70'),#R23
    
    #4
    ('~C3',),
    ('~C75',),#R24
    ('A3','B3'),#R25
    ('C4','C2'),#R26
    ('~C70','A75','B75'),#R27
    
    #5
    ('~A2',),#R28
    ('A3', 'A4'),#R29
    ('~B2','C3','C4'),#R30
    ('~B2','A4','A3'),#R31
    ('~C2','B3','B4'),#R32
    ('~C2','A4'),#R33
    
    #Constraints
    ('AH', 'AG', 'AF'), 
    ('~AH', '~AG'), 
    ('~AH', '~AF'), 
    ('~AF', '~AG'),
    ('BH', 'BG', 'BF'), 
    ('~BH', '~BG'), 
    ('~BH', '~BF'), 
    ('~BF', '~BG'),
    ('CH', 'CG', 'CF'), 
    ('~CH', '~CG'), 
    ('~CH', '~CF'), 
    ('~CF', '~CG'),
    ('A65', 'A70', 'A75'), 
    ('~A65', '~A70'), 
    ('~A65', '~A75'), 
    ('~A75', '~A70'),
    ('B65', 'B70', 'B75'), 
    ('~B65', '~B70'), 
    ('~B65', '~B75'), 
    ('~B75', '~B70'),
    ('C65', 'C70', 'C75'), 
    ('~C65', '~C70'), 
    ('~C65', '~C75'), 
    ('~C75', '~C70'),
    ('A2', 'A3', 'A4'), 
    ('~A2', '~A3'), 
    ('~A2', '~A4'), 
    ('~A3', '~A4'),
    ('B2', 'B3', 'B4'), 
    ('~B2', '~B3'), 
    ('~B2', '~B4'), 
    ('~B3', '~B4'),
    ('C2', 'C3', 'C4'), 
    ('~C2', '~C3'), 
    ('~C2', '~C4'), 
    ('~C3', '~C4'),
]

Symbols = (
    'AF', 'AG', 'AH', 'BF', 'BG', 'BH', 'CF', 'CG', 'CH', 
    'A65', 'A70', 'A75', 'B65', 'B70', 'B75', 
    'C65', 'C70', 'C75', 'A2', 'A3', 'A4', 
    'B2', 'B3', 'B4', 'C2', 'C3', 'C4'
)
# model = {i: None for i in Symbols}
# _, unique_model = BackTrackingSearch(Clauses, model)
# print(unique_model)

model = {i: None for i in Symbols}
_, model = BackTrackingSearch(Clauses, model)
BackTrackingSearch(Clauses, model)

(True,
 {'AF': False,
  'AG': False,
  'AH': True,
  'BF': True,
  'BG': False,
  'BH': False,
  'CF': False,
  'CG': True,
  'CH': False,
  'A65': True,
  'A70': False,
  'A75': False,
  'B65': False,
  'B70': False,
  'B75': True,
  'C65': False,
  'C70': True,
  'C75': False,
  'A2': False,
  'A3': False,
  'A4': True,
  'B2': False,
  'B3': True,
  'B4': False,
  'C2': True,
  'C3': False,
  'C4': False})

### Hence a model that satisfies the knowledge base is derived from above.

## c. Determining uniqueness of the KB

In [8]:
T=True
F=False

model = [T,F,T,T,F,T,T,F,T,F,T,T,F,F,T,T,F,T,F,T,T,T,F,F,T,F,T,F,F,T,T,F,F,T,T,T,F,T,F,F,F,F,T,T,F,T,F,T,T,T,F]

In [6]:

implies= lambda P,Q: (not P) or Q
Constraint = lambda P,Q,R: (P and not Q and not R) or ( not P and Q and not R) or (not P and not Q and R)

KnowledgeBase =[
    lambda model: not model[2] and not model[26],# R1
    lambda model: model[5] or model[8],#R2
    lambda model: implies((model[0]),(model[4] or model[7])), #R3
    lambda model: implies((model[1]),(model[3] or model[6])), #R4
    lambda model: not model[7] or not model[10],#R5
    lambda model: model[4] or model[1],#R6
    lambda model: model[13] or model[16],#R7
    lambda model: model[6] or model[8],#R8
    lambda model: model[9] or model[11],#R9
    lambda model: implies((model[6]),(model[11])),#R10
    lambda model: implies((model[6]),(model[12] or model[15])),#R11
    lambda model: implies((model[6]),(model[5] or model[2])),#R12
    lambda model: implies((model[8]),(model[14] or model[17])),#R13
    lambda model: implies((model[8]),(model[3] or model[0])),#R14
    lambda model: implies((model[11]),(model[12] or model[15])),#R15
    lambda model: implies((model[9]),(model[14] or model[17])),#R16
    lambda model: implies((model[5]),(model[23])),#R17
    lambda model: implies((model[2]),(model[23])),#R18
    lambda model: implies((model[1]),(model[25])),#R19
    lambda model: implies((model[0]),(model[24])),#R20
    lambda model: implies((model[8]),(model[20])),#R21
    lambda model: implies((model[8]),(model[1] or model[22])),#R22
    lambda model: implies((model[8]),(model[0] or model[21])),#R23
    lambda model: not model[14] and not model[26],#R24
    lambda model: model[12] or model[13],#R25
    lambda model: model[17] or model[11],#R26
    lambda model: implies((model[23]),(model[24] or model[25])),#R27
    lambda model: not model[9],#R28
    lambda model: model[12] or model[15],#R29
    lambda model: implies((model[10]),(model[14] or model[17])),#R30
    lambda model: implies((model[10]),(model[12] or model[15])),#R31
    lambda model: implies((model[11]),(model[13] or model[16])),#R32
    lambda model: implies((model[11]),(model[15])),#R33
    lambda model : Constraint(model[0],model[1],model[2]), # R34
    lambda model : Constraint(model[3],model[4],model[5]), # R35
    lambda model : Constraint(model[6],model[7],model[8]), # R36
    lambda model : Constraint(model[9],model[10],model[11]), # R37
    lambda model : Constraint(model[12],model[13],model[14]), # R38
    lambda model : Constraint(model[15],model[16],model[17]), # R39
    lambda model : Constraint(model[18],model[19],model[20]), # R40
    lambda model : Constraint(model[21],model[22],model[23]), # R41
    lambda model : Constraint(model[24],model[25],model[26]), # R142
    lambda model : Constraint(model[0],model[3],model[6]), # R43
    lambda model : Constraint(model[1],model[4],model[7]), # R44
    lambda model : Constraint(model[2],model[5],model[8]), # R45
    lambda model : Constraint(model[9],model[12],model[15]), # R46
    lambda model : Constraint(model[10],model[13],model[16]), # R47
    lambda model : Constraint(model[11],model[14],model[17]), # R48
    lambda model : Constraint(model[18],model[21],model[24]), # R49
    lambda model : Constraint(model[19],model[22],model[25]), # R50
    lambda model : Constraint(model[20],model[23],model[26]), # R51
                ]
   
import functools
def evaluateKB(Model, KnowledgeBase):
    Result = [KB(Model) for KB in KnowledgeBase]  # Evaluate each sentence in the knowledge base for the model
    TrueFalse = functools.reduce(lambda a, b: a and b, Result)  # Use a reduce operation to compute the logical AND of all sentence truth values
    return TrueFalse, Result

import itertools
TrueModelCount = 0
Models = itertools.product([True,False], repeat = 27)
for model in Models:
    TF, Result = evaluateKB(model,KnowledgeBase)
    if TF:
        TrueModelCount+=1
        print(f"---\nTrue Model {TrueModelCount}: {model}")
        print(f"Result: {Result}")
print(f"===\nThere are {TrueModelCount} models for which the KB is True")

---
True Model 1: (False, True, False, False, False, True, True, False, False, False, False, True, False, True, False, True, False, False, True, False, False, False, False, True, False, True, False)
Result: [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
===
There are 1 models for which the KB is True


### Hence the model is uniquely satisfiable.

## d. Improving the backtracking search method by eliminating “trivial solutions” entailed by so-called unit clauses.

In [13]:
def preprocess_unit_clauses(Clauses):
    unit_literals = set()
    
    for clause in Clauses:
        if len(clause) == 1:
            unit_literals.add(clause[0])

    model = {}
    for literal in unit_literals:
        if literal[0] == '~':
            model[literal[1:]] = False
        else:
            model[literal] = True

    return model


initial_model = preprocess_unit_clauses(Clauses)
print("Initial Model from Unit Clauses:", initial_model)

# Merge the initial model with the model used in BackTrackingSearch
model.update(initial_model)

# Run the BackTrackingSearch with the merged model
final_model = BackTrackingSearch(Clauses, model)

# Print the final model
print("Final Model:", final_model)

Initial Model from Unit Clauses: {'C3': False, 'C75': False, 'CF': False, 'A2': False}
Final Model: (True, {'AF': False, 'AG': False, 'AH': True, 'BF': True, 'BG': False, 'BH': False, 'CF': False, 'CG': True, 'CH': False, 'A65': True, 'A70': False, 'A75': False, 'B65': False, 'B70': False, 'B75': True, 'C65': False, 'C70': True, 'C75': False, 'A2': False, 'A3': False, 'A4': True, 'B2': False, 'B3': True, 'B4': False, 'C2': True, 'C3': False, 'C4': False}, 24)


## e. Number of models  evaluated before a solution was found.

###  Evaluation of number of models for backtracking search (b)

In [5]:
import copy

def EvaluateKB(Clauses, Model):
    EvaluatedClauses = []
    for clause in Clauses:
        EvaluatedLiterals = []
        for literal in clause:
            if literal[0] == '~':
                negated_literal = literal[1:]
                if Model.get(negated_literal) is None:
                    EvaluatedLiterals.append(None)
                else:
                    EvaluatedLiterals.append(not Model[negated_literal])
            else:
                if Model.get(literal) is None:
                    EvaluatedLiterals.append(None)
                else:
                    EvaluatedLiterals.append(Model[literal])
        if True in EvaluatedLiterals:
            EvaluatedClauses.append(True)
        elif None in EvaluatedLiterals:
            EvaluatedClauses.append(None)
        else:
            EvaluatedClauses.append(False)

    if all(i is True for i in EvaluatedClauses):
        EvaluatedKB = True
    elif False in EvaluatedClauses:
        EvaluatedKB = False
    else:
        EvaluatedKB = None

    return EvaluatedKB
def BackTrackingSearch(Clauses, Model, evaluation_count=0):
    EvaluatedClauses = EvaluateKB(Clauses, Model)
    evaluation_count += 1  # Increment the evaluation count

    if EvaluatedClauses is True:
        return True, Model, evaluation_count
    elif EvaluatedClauses is False:
        return False, None, evaluation_count
    else:
        Model2 = copy.deepcopy(Model)
        NextSymbol = [i for i in Model2 if Model2[i] is None][0]
        Model2[NextSymbol] = True
        ModelTrue, SATModel, count_true = BackTrackingSearch(Clauses, Model2, evaluation_count)
        if ModelTrue:
            return True, SATModel, count_true

        Model2[NextSymbol] = False
        ModelFalse, SATModel, count_false = BackTrackingSearch(Clauses, Model2, evaluation_count)
        if ModelFalse:
            return True, SATModel, count_false

    return False, None, evaluation_count
Clauses = [
   #1
    ('~CF',),
    ('~C75',),#R1
    ('CG', 'CH'),#R2
    ('~AF','BG','BH'),#R3
    ('~BF','AG','AH'),#R4
    #2
    ('~BH', '~B2'),#R5
    ('BG','BF'),#R6
    ('B3','B4'),#R7
    ('AH','CH'),#R8
    ('A2','C2'),#R9
    ('~AH', 'C2'),#R10
    ('~AH','A3','A4'),#R11
    ('~AH','CG','CF'),#R12
    ('~CH','C3','C4'),#R13
    ('~CH','AG','AF'),#R14
    ('~C2', 'A3','A4'),#R15
    ('~A2','C3','C4'),#R16
   
    #3
    ('~CG','C70'),#R17
    ('~CF','C70'),#R18
    ('~BF','B75'),#R19
    ('~AF','A75'),#R20
    ('~CH','C65'),#R21
    ('~CH','BF','B70'),#R22
    ('~CH','AF','A70'),#R23
    
    #4
    ('~C3',),
    ('~C75',),#R24
    ('A3','B3'),#R25
    ('C4','C2'),#R26
    ('~C70','A75','B75'),#R27
    
    #5
    ('~A2',),#R28
    ('A3', 'A4'),#R29
    ('~B2','C3','C4'),#R30
    ('~B2','A4','A3'),#R31
    ('~C2','B3','B4'),#R32
    ('~C2','A4'),#R33
    
    #Constraints
    ('AH', 'AG', 'AF'), 
    ('~AH', '~AG'), 
    ('~AH', '~AF'), 
    ('~AF', '~AG'),
    ('BH', 'BG', 'BF'), 
    ('~BH', '~BG'), 
    ('~BH', '~BF'), 
    ('~BF', '~BG'),
    ('CH', 'CG', 'CF'), 
    ('~CH', '~CG'), 
    ('~CH', '~CF'), 
    ('~CF', '~CG'),
    ('A65', 'A70', 'A75'), 
    ('~A65', '~A70'), 
    ('~A65', '~A75'), 
    ('~A75', '~A70'),
    ('B65', 'B70', 'B75'), 
    ('~B65', '~B70'), 
    ('~B65', '~B75'), 
    ('~B75', '~B70'),
    ('C65', 'C70', 'C75'), 
    ('~C65', '~C70'), 
    ('~C65', '~C75'), 
    ('~C75', '~C70'),
    ('A2', 'A3', 'A4'), 
    ('~A2', '~A3'), 
    ('~A2', '~A4'), 
    ('~A3', '~A4'),
    ('B2', 'B3', 'B4'), 
    ('~B2', '~B3'), 
    ('~B2', '~B4'), 
    ('~B3', '~B4'),
    ('C2', 'C3', 'C4'), 
    ('~C2', '~C3'), 
    ('~C2', '~C4'), 
    ('~C3', '~C4'),
]

Symbols = (
    'AF', 'AG', 'AH', 'BF', 'BG', 'BH', 'CF', 'CG', 'CH', 
    'A65', 'A70', 'A75', 'B65', 'B70', 'B75', 
    'C65', 'C70', 'C75', 'A2', 'A3', 'A4', 
    'B2', 'B3', 'B4', 'C2', 'C3', 'C4'
)

model = {i: None for i in Symbols}
_, final_model, evaluation_count = BackTrackingSearch(Clauses, model)
print("Final Model:", final_model)
print("Number of Models Evaluated:", evaluation_count)

Final Model: {'AF': False, 'AG': False, 'AH': True, 'BF': True, 'BG': False, 'BH': False, 'CF': False, 'CG': True, 'CH': False, 'A65': True, 'A70': False, 'A75': False, 'B65': False, 'B70': False, 'B75': True, 'C65': False, 'C70': True, 'C75': False, 'A2': False, 'A3': False, 'A4': True, 'B2': False, 'B3': True, 'B4': False, 'C2': True, 'C3': False, 'C4': False}
Number of Models Evaluated: 28


### Here, the final result is obtained after the backtracking search is done on 28 models. 

### Evaluation of models for uniquely satisfiable model (c)

In [7]:

T = True
F = False

model = [T, F, T, T, F, T, T, F, T, F, T, T, F, F, T, T, F, T, F, T, T, T, F, F, T, F, T, F, F, T, T, F, F, T, T, T, F, T, F, F, F, F, T, T, F, T, F, T, T, T, F]

implies = lambda P, Q: (not P) or Q
Constraint = lambda P, Q, R: (P and not Q and not R) or (not P and Q and not R) or (not P and not Q and R)

KnowledgeBase = [
    lambda model: not model[2] and not model[26],# R1
    lambda model: model[5] or model[8],#R2
    lambda model: implies((model[0]),(model[4] or model[7])), #R3
    lambda model: implies((model[1]),(model[3] or model[6])), #R4
    lambda model: not model[7] or not model[10],#R5
    lambda model: model[4] or model[1],#R6
    lambda model: model[13] or model[16],#R7
    lambda model: model[6] or model[8],#R8
    lambda model: model[9] or model[11],#R9
    lambda model: implies((model[6]),(model[11])),#R10
    lambda model: implies((model[6]),(model[12] or model[15])),#R11
    lambda model: implies((model[6]),(model[5] or model[2])),#R12
    lambda model: implies((model[8]),(model[14] or model[17])),#R13
    lambda model: implies((model[8]),(model[3] or model[0])),#R14
    lambda model: implies((model[11]),(model[12] or model[15])),#R15
    lambda model: implies((model[9]),(model[14] or model[17])),#R16
    lambda model: implies((model[5]),(model[23])),#R17
    lambda model: implies((model[2]),(model[23])),#R18
    lambda model: implies((model[1]),(model[25])),#R19
    lambda model: implies((model[0]),(model[24])),#R20
    lambda model: implies((model[8]),(model[20])),#R21
    lambda model: implies((model[8]),(model[1] or model[22])),#R22
    lambda model: implies((model[8]),(model[0] or model[21])),#R23
    lambda model: not model[14] and not model[26],#R24
    lambda model: model[12] or model[13],#R25
    lambda model: model[17] or model[11],#R26
    lambda model: implies((model[23]),(model[24] or model[25])),#R27
    lambda model: not model[9],#R28
    lambda model: model[12] or model[15],#R29
    lambda model: implies((model[10]),(model[14] or model[17])),#R30
    lambda model: implies((model[10]),(model[12] or model[15])),#R31
    lambda model: implies((model[11]),(model[13] or model[16])),#R32
    lambda model: implies((model[11]),(model[15])),#R33
    lambda model : Constraint(model[0],model[1],model[2]), # R34
    lambda model : Constraint(model[3],model[4],model[5]), # R35
    lambda model : Constraint(model[6],model[7],model[8]), # R36
    lambda model : Constraint(model[9],model[10],model[11]), # R37
    lambda model : Constraint(model[12],model[13],model[14]), # R38
    lambda model : Constraint(model[15],model[16],model[17]), # R39
    lambda model : Constraint(model[18],model[19],model[20]), # R40
    lambda model : Constraint(model[21],model[22],model[23]), # R41
    lambda model : Constraint(model[24],model[25],model[26]), # R142
    lambda model : Constraint(model[0],model[3],model[6]), # R43
    lambda model : Constraint(model[1],model[4],model[7]), # R44
    lambda model : Constraint(model[2],model[5],model[8]), # R45
    lambda model : Constraint(model[9],model[12],model[15]), # R46
    lambda model : Constraint(model[10],model[13],model[16]), # R47
    lambda model : Constraint(model[11],model[14],model[17]), # R48
    lambda model : Constraint(model[18],model[21],model[24]), # R49
    lambda model : Constraint(model[19],model[22],model[25]), # R50
    lambda model : Constraint(model[20],model[23],model[26]), # R51
                ]

import functools

def evaluateKB(Model, KnowledgeBase, counter):
    Result = [KB(Model) for KB in KnowledgeBase]  # Evaluate each sentence in the knowledge base for the model
    TrueFalse = functools.reduce(lambda a, b: a and b, Result)  # Use a reduce operation to compute the logical AND of all sentence truth values
    counter[0] += 1  # Increment the counter
    return TrueFalse, Result

import itertools

TrueModelCount = 0
Models = itertools.product([T, F], repeat=27)
counter = [0]  # Counter to keep track of the number of models evaluated

for model in Models:
    TF, Result = evaluateKB(model, KnowledgeBase, counter)
    if TF:
        TrueModelCount += 1
        print(f"---\nTrue Model {TrueModelCount}: {model}")
        print(f"Result: {Result}")
        print(f"Models evaluated before solution: {counter[0]}")
        break  # Stop after finding the first true model

print(f"===\nThere are {TrueModelCount} models for which the KB is True")

---
True Model 1: (False, True, False, False, False, True, True, False, False, False, False, True, False, True, False, True, False, False, True, False, False, False, False, True, False, True, False)
Result: [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
Models evaluated before solution: 97474294
===
There are 1 models for which the KB is True


### The number of models evaluated before finding 1 true model is 97474294 models.

### Evaluating number of models for improved backtracking search (d)

In [28]:
def preprocess_unit_clauses(Clauses):
    unit_literals = set()

    for clause in Clauses:
        if len(clause) == 1:
            unit_literals.add(clause[0])

    model = {}
    for literal in unit_literals:
        if literal[0] == '~':
            model[literal[1:]] = False
        else:
            model[literal] = True

    return model

initial_model = preprocess_unit_clauses(Clauses)
print("Initial Model from Unit Clauses:", initial_model)

# Merge the initial model with the model used in BackTrackingSearch
model_copy = dict(model)  # Make a copy of the original model
model_copy.update(initial_model)

# Run the BackTrackingSearch with the merged model
result_tuple = BackTrackingSearch(Clauses, model_copy)
result, final_model, evaluation_count = result_tuple

# Print the final model and the number of models evaluated
print("Final Model:", final_model)
print("Number of models evaluated:", evaluation_count)

Initial Model from Unit Clauses: {'C3': False, 'C75': False, 'CF': False, 'A2': False}
Final Model: {'AF': False, 'AG': False, 'AH': True, 'BF': True, 'BG': False, 'BH': False, 'CF': False, 'CG': True, 'CH': False, 'A65': True, 'A70': False, 'A75': False, 'B65': False, 'B70': False, 'B75': True, 'C65': False, 'C70': True, 'C75': False, 'A2': False, 'A3': False, 'A4': True, 'B2': False, 'B3': True, 'B4': False, 'C2': True, 'C3': False, 'C4': False}
Number of models evaluated: 24


### Here, the number of models evaluated for the improved backtracking search is 24. This is the expected result as the number of evaluated models is less than backtracking without tuning.
